File: naddu/naddu_2/c/move.c
#include "move.h"

uint32_t move_is_probably_legal(Node *node, uint32_t move) {
  if (move == 0)
    return 0;

  Pos *pos = &node->pos;
  uint8_t *board = pos->board;
  int stm = pos->stm;

  int from = MOVE_FROM(move);
  int to = MOVE_TO(move);

  // from and to must be on board
  if ((from | to) & 0x88)
    return 0;

  // must have our piece on from square
  int piece = board[from];
  if (!piece)
    return 0;
  if ((piece & BLACK) != stm)
    return 0;

  // to square must be empty or enemy piece
  int target = board[to];
  if (target && (target & BLACK) == stm)
    return 0;

  return move;
}

void format_move(uint32_t move, char *buf) {
  int from = MOVE_FROM(move);
  int to = MOVE_TO(move);

  buf[0] = 'a' + (from & 7);
  buf[1] = '1' + (from >> 4);
  buf[2] = 'a' + (to & 7);
  buf[3] = '1' + (to >> 4);

  if (move & MOVE_PROMO_MASK) {
    int promo_type = (move & MOVE_PROMO_MASK) >> MOVE_PROMO_SHIFT;
    const char promo_chars[] = {0, 0, 'n', 'b', 'r', 'q'};
    buf[4] = promo_chars[promo_type];
    buf[5] = '\0';
  }
  else {
    buf[4] = '\0';
  }
}

File: naddu/naddu_2/src/move.js
function moveIsProbablyLegal(node, move) {

  if (move === 0)
    return 0;
  
  const pos = node.pos;
  const board = pos.board;
  const stm = pos.stm;

  const from = (move >> 8) & 0xff;
  const to = move & 0xff;

  // from and to must be on board
  if ((from | to) & 0x88) 
    return 0;

  // must have our piece on from square
  const piece = board[from];
  if (!piece) 
    return 0;
  if ((piece & BLACK) !== stm) 
    return 0;

  // to square must be empty or enemy piece
  const target = board[to];
  if (target && (target & BLACK) === stm) 
    return 0;

  return move;
}

function formatMove(move) {

  const from = (move >> 8) & 0xff;
  const to = move & 0xff;

  const fromFile = String.fromCharCode(97 + (from & 7));
  const fromRank = String.fromCharCode(49 + (from >> 4));
  const toFile = String.fromCharCode(97 + (to & 7));
  const toRank = String.fromCharCode(49 + (to >> 4));

  let moveStr = fromFile + fromRank + toFile + toRank;

  // Add promotion piece if applicable
  if (move & MOVE_PROMO_MASK) {
    const promoType = (move & MOVE_PROMO_MASK) >> MOVE_PROMO_SHIFT;
    const promoChars = ['', '', 'n', 'b', 'r', 'q'];
    moveStr += promoChars[promoType];
  }

  return moveStr;
}
